### ISSUE TITLE
Improve the search to be able to find anything that we map

### ISSUE THREAD

**Issue by zstadler** [IsraelHikingMap/Site#132]
Adding the ability to search names of trails

**Comment by HarelM** [comment]
Could be an easy task if you want to try it out for yourself:
the search controller is using the Nominatim API.
the method name is: ```private handleSearchResponse``` in ```SearchController ```.
If not, I'll add it myself :-)
if you have a trail name that should come out when searching let me know so I I'll know if the implementation is correct.

**Comment by zstadler** [comment]
It seems like the Nominatim API does not search relations.

For example, [מים אל ים relation ](http://www.openstreetmap.org/relation/2860967) cannot be found by the API.

The [Mapquest Nominatim API](http://open.mapquestapi.com/nominatim/) may be able to support search on relations. However, it required a developer key, available [here](https://developer.mapquest.com/).

**Comment by HarelM** [comment]
I'm thinking about using the following project:
https://github.com/komoot/photon instead of the limited nominatim.
I'm still reading the documentation in order to understand what I need to do in order to get it looking into the pbf file of Israel...

**Comment by HarelM** [comment]
Photon is based on the Nominatim database and does not have relations as far as I can tell...
Issue can be tracked here: https://github.com/twain47/Nominatim/issues/413

**Comment by zstadler** [comment]
Is the search confined to relations with a "route" tag?

**Comment by HarelM** [comment]
Nope, the relation search uses overpass rel["name"~<search term>] other features are fetched from nominatim.

**Comment by zstadler** [comment]
It looks like the Overpass-Turbo search is using the first word of the search term.

For example, searching for "נחל נקרות" shows many relations, all of them have "נחל" in their names.
On the other hand, searching for "נקרות" shows no relations, as it should

**Comment by HarelM** [comment]
I'm working on creating an [elastic search](https://www.elastic.co/products/elasticsearch) database with all relevant data on our server and moving the search to be entirely from there, no outside source. I hope it will yield good results...

**Comment by HarelM** [comment]
Ok, Managed to find a [library](http://www.osmsharp.com/) to read the osm pbf file.
Converted all the data to geojson and indexed it to the elastic search data base.
Now when I search I get a lot of results including anything that has a name in the OSM database.
In order to make the search better I need to be able to sort it according to tag priority.
@zstadler @valleyofdawn Can you help me in the above task? Can you prioritize tags and values in this thread?
Let's see if low, normal and high will be enough, if not we can use numbers (1 to 10 for example).
and example of priority is:
**relation: high
highway:residential is low.**
I'm going to present the user with 5-10 search results so if there are many results this field becomes important.

**Comment by zstadler** [comment]
I think we should exclude some categories first, based on the actual results. For example, there are ~30,000 highway=bus_stop nodes with name.

**Comment by HarelM** [comment]
Don't worry about the number of results, current script to open the pbf file and insert all 130,000 records takes about 1 minute, search is very fast even with all that data. making **bus_stop** as **medium** priority should do the trick and if there are more relevant matching they will be presented first.

**Comment by zstadler** [comment]
Is it possible to have everything in medium priority and select specific tags as either high or low?

IMHO, High priority, in no particular order:
- relation boundary=national_park 
- relation boundary=protected_area 
- relation leisure=nature_reserve
- relation @ @isOneOf(route, hiking, bicycle, mtb)
- way [highway] AND [highway!="construction"]
- way  @isOneOf(waterway, stream, river, wadi)
- way boundary=national_park 
- way boundary=protected_area 
- way leisure=nature_reserve
- node place 
- node landuse=farmyard
- node @isOneOf(natural, peak, spring, tree, cave_entrance, waterhole) 
- node water=pond 
- node @isOneOf(man_made, water_well, cistern) 
- node @isOneOf(leisure, picnic, picnic_table) 
- node @isOneOf(tourism, picnic_site, camp_site, viewpoint, attraction) 
- node @isOneOf(historic, ruins, archaeological_site, memorial, monument)

**Comment by HarelM** [comment]
Done.
@zstadler @valleyofdawn please help me fine-tune the search now that it is working with our local elastic search service.
I'll leave this issue open for feedback for the next week.

**Comment by zstadler** [comment]
Low priority:
- node place=suburb

**Comment by HarelM** [comment]
Current results are below:
I'm digging into the documentation and I found out the following:
Search results have a score between 0.0 and 1.0 the higher the score the better the results according to ES score function.
I have added another sorting so that all the high features have 1 and the other have 0 and the current sort is first by this field first and then by score.
this means that if you search for a feature that has a perfect score (perfect match) but is not in our list it might get voted down.
I'm not sure I know how to address it or how to improve the search further, thought are welcome.
I don't know if I can some the score with another field but the documentation does not describe such an option...  
![image](https://cloud.githubusercontent.com/assets/3269297/14124544/c1ba9d54-f60e-11e5-9511-7fba324db668.png)

**Comment by zstadler** [comment]
According to the [Multipolygon Usage rules](http://wiki.openstreetmap.org/wiki/Relation:multipolygon#Usage), the relation's members should remain untagged, with some exceptions.

Currently, the search fails on such cases. For example, the search for "[יער יתיר](http://www.openstreetmap.org/relation/1653352)" - a relation with 7 untagged member ways, is not fouund by the search.

**Comment by HarelM** [comment]
This is not how I understand it, but I might be misinterpreting.
"In short, a multipolygon relation can have any number of ways in the role outer (the outline) and any number of ways in the role inner (the holes), and these must somehow form valid rings to build a multipolygon from".

Regarding forests, they are not in the list in the file:
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.API/Services/GeoJsonFeatureHelper.cs
Do we want to add them?

**Comment by zstadler** [comment]
I real the quoted paragraph as describing the required topology of a multipolygon. I don't see how it relates to the search, as it does not consider the shapes.

**Comment by zstadler** [comment]
When a relation is selected in the search results, all members should be shown on the map and the map zoomed-to all of them.
Similarly, when a way is selected in the search results, the map should zoom-to all of its extent.

Since Israeli forest relations tend to have many small segments, especially in the south, forests should be added to the search only after the search selection is changed as suggested above.

**Comment by HarelM** [comment]
I agree about the zooming.
The reason the search is sensitive to multipolygon's members roles is that it needs the roles in order to convert the data to geojson to save in elastic search database. When the shape is not well defined the conversion might fail, I fixed the current conversion to be a bit more robust.
Regarding search results: exact match is in second priority to current elements list defined above. I couldn't find a way to merge the two calculation together, if you have so time to go over the elastic search documentation to look for a better solution please let me know.

**Comment by HarelM** [comment]
Search should be better now (See check-in comments). See if find something out of order...

**Comment by zstadler** [comment]
This is definitely an improvement! 
Search for "שמורת טבע מצוק הצינים" works well now.

Search for "גן לאומי בית-שאן" fails to find [way 394429011](http://www.openstreetmap.org/way/394429011) or any of the other outer members of [relation 5917822](http://www.openstreetmap.org/relation/5917822).

Searching for "נחל שילה" finds 9 different segments of the stream.

When street is found, it would be nice to give the city/locality. For example "המכתש".

It looks like I have expected from the elastic search more than it can deliver. Searching for "נחל שלה" does not find "נחל שילה". Similarly, search for "המכתש" and "מכתש".

**Comment by HarelM** [comment]
Thanks for the thorough tests, this really helps me find the issues.
Some issues I found in בית שאן:
There is a relation without a name holding a lot of way with the name גן לאומי בית שאן.
There is a relation that is not multipolygon that holds a polygon and not a regular way, not sure it is legal, but now the code knows how to handle it...
My current logic in order to avoid duplication does the following:
1. Looks for relations that has ways with the same name and removed the ways from the results.
2. Looks for places with the same name for way and node or way and relation and removes the node.
Keep up looking for places that do not get to the search results and I'll investigate what caused them to disappear.

**Comment by zstadler** [comment]
Indeed the current tagging of "boundary=protected_area" relations and ways is incompatible with OSM guidelines. The members of the relation should not duplicate the tags of the relation. I'm fixing this "as we speak". After I'm done, there should not be such duplication in the OSM data. 

It is good to have an fail-safe mechanism in the search as this mistake can be made again in the future...


Note that "type=boundary" and "boundary=*" relations should be treated the same as "type=multipolygon"

**Comment by HarelM** [comment]
Currently anything that is not multipolygon is considered as mutliline in terms of what geographic shape is stored in the search database. Since this does not affect anything at the moment I don't think I should add workaround to fix the shape it self - search results are still marked on the map with a blue line.

**Comment by zstadler** [comment]
Does the search treat "type=boundary" and "boundary=*" relations as multipolygos or not?
I think it should.

**Comment by zstadler** [comment]
When selecting a POI from the search, the default zoom is 20. I suggest the minimal zoom for focusing on a search result should be 16. This is also the current maximal zoom level of the map.

**Comment by HarelM** [comment]
All relations are converted to either multiploygon or multiline. it term of correct behavior code-wise nothing that is not a ```type=multipoligon``` should not be converted to multipolygon.
Again, since this is not visible by the user, I don't see how this is an issue currently.
I agree about the zoom 20, this is the default behavior, I saw it and forgot to fix it.

**Comment by zstadler** [comment]
What then is the purpose of creating multipolygons if the difference is not visible?
Alternatively, should there be a visible difference for multipolygons, such as a transparent fill color?

**Comment by HarelM** [comment]
Well, the purpose was to write a converter between OSM data and GeoJSON since OSM data is in XML and can't be saved into the elastic search database. It's up to us if we want to fill multipolygone and polygon shapes.
I don't think that currently we want to support "bad" mapping - in terms of showing shapes that are not multipoligon as multipoligon.

**Comment by zstadler** [comment]
This is the opposite of bad mapping.
According to the [wiki on boundary relations](http://wiki.openstreetmap.org/wiki/Relation:boundary), administrative, protected_area, and national_park boundary relations are areas with enclaves/exclaves, just like multipolygons. The use of type=multipolygon tag on boundaries is deprecated.

**Comment by HarelM** [comment]
I see now. I'll fix it then. Thanks for the info.

**Comment by zstadler** [comment]
Thanks!
As you wrote, it does not seem to have any user-visible impact (yet).

**Comment by zstadler** [comment]
Can we index all *_name, name:* and name[1-9] tags too?
We may be missing some multi-lingual and alternative names.

**Comment by HarelM** [comment]
Every name gets indexed, only "name" and "name:en" are being searched according to the user typed language.
I'll see if I can find something in the elastic search documentation how to perform this, but I assume it won't be easy/possible...

**Comment by HarelM** [comment]
Added support for _name and name*.
Added support for fuzzyness - נחל שלה now works.
Fixed search score so that search calculates the number we add as factor - in other words I incorporated the "important" points in the score calculation.
Fixed zoom of point results.
Still to be adjusted: the names of the street to say to which city they belong.

**Comment by zstadler** [comment]
Currently, each segment of a road is a separate result. For example, "קריית שאול".

**Comment by zstadler** [comment]
Do we want to include search results from Wikipedia? 
Only when that Wiki layer is shown?

**Comment by HarelM** [comment]
As far as I can tell you can only search wiki by location in a radius of 10Km, so searching it from the search bar is not really an option.

**Comment by HarelM** [comment]
I'll see if I can make the grouping logic a bit better in order to combine roads and streams that have the same name and have a point where they join.

**Comment by HarelM** [comment]
You can check-out נחל שילה now, it should only contain one search result.
Notice that קריית שאול is written קרית שאול in some places and can't be merged out of the box...

**Comment by zstadler** [comment]
Excellent!
I've unified the spelling of "קריית שאול" so we'll be able to check it in a couple of days. 
Note that the topology is complex and will probably need a multi-polyline for its representation.

**Comment by HarelM** [comment]
It didn't work for קריית שאול... :-(
I'll think about it, but I guess that for this particular case since it is a street that contains multiple line it's probably better to have a relation related to it instead of duplicating the name so many time - this way you don't have spelling mistakes too :-)

**Comment by zstadler** [comment]
I didn't see a use case of creating relations for streets. Perhaps I can
split and merge the segments differently to reduce the number of separate
results.

**Comment by HarelM** [comment]
Closing this issue now as the major development has been completed. allowing other issues to be open according to specific problems.
